src/lib/agent.ts

```diff
@@ -62,13 +69,71 @@ export class Agent {
     eventbus.on("didGenerated", callback)
   }
 
+  private handleDiscoverFeatures(message: IMessage) {
+    const regexEscape = (s: string) => s.replace(/([.*+?$^=!:{}()|\[\]\/\\])/g, "\\$1");
+    const createRegex = (query: string) => (new RegExp(`^${query.split("*").map(regexEscape).join(".*")}$`));
+    let protocolResponse: object[] = [];
+
+    // Loop through all queries, then all implemented protocols and build up a
+    // list of supported protocols that match the user's request
+    for(let query of message.body.queries) {
+
+      // Rudimentary implementation, ignoring all except protocol requests
+      if(query["feature-type"] != "protocol")
+        continue
+
+      for(let protocol of IMPLEMENTED_MESSAGE_TYPES) {
+        if(createRegex(query["match"]).test(protocol)) {
+          protocolResponse.push({
+            "feature-type": "protocol",
+            "id": protocol,
+          })
+        }
+      }
+    }
+    const response = {
+      "type": "https://didcomm.org/discover-features/2.0/disclose",
+      "thid": message.id,
+      "body": {
+        "disclosures": protocolResponse
+      }
+    }
+    return response
+  }
+
+  private handleCoreProtocolMessage(message: IMessage) {
+    switch(message.type) {
+      case "https://didcomm.org/trust-ping/2.0/ping":
+        if(message.body?.response_requested !== false) {
+          this.sendMessage({
+              did: message.from
+            }, {
+            "type": "https://didcomm.org/trust-ping/2.0/ping-response",
+            "thid": message.id,
+          })
+        }
+        break;
+      case "https://didcomm.org/discover-features/2.0/queries":
+        const discloseMessage = this.handleDiscoverFeatures(message)
+        this.sendMessage({
+              did: message.from
+            }, discloseMessage)
+        break;
+    }
+  }
+
   private onMessageReceived(message: IMessage) {
     const from = message.from == this.profile.did ? this.profile as Contact : ContactService.getContact(message.from)
     const to = message.to[0] == this.profile.did ? this.profile as Contact : ContactService.getContact(message.to[0])
+
+    // Handle core protocols first, but only if they are destined for us. Don't
+    // handle them if they were sent by us.
+    if(this.profile && to?.did == this.profile?.did)
+      this.handleCoreProtocolMessage(message)
```

Comment:
When we say a message sent by us, do we mean a message we literally posted to the mediator for ourselves? `onMessageReceived` should only be messages we're receiving rather than sending, right?

Reply:
You're probably right about that one :sweat_smile: